cmd/locateCG/locateCG.go

```diff
@@ -0,0 +1,193 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fasta"
+	"github.com/vertgenlab/gonomics/fileio"
+)
+
+type Settings struct {
+	InFa      string //input fasta file with 2 aligned sequences to compare
+	ChromName string //e.g. chr1
+}
+
+type CGsite struct {
+	Chrom    string //derived from ChromName
+	StartPos int    //start pos of CG site in reference genome of sequence 1
+	EndPos   int    //end pos of CG site in reference genome of sequence 1
+	Type     string //gain or loss in sequence 1 relative to sequence 2
+	AlnStart int    //corresponding alignment position of reference StartPos
+	AlnEnd   int    //corresponding alignment position of reference EndPos
+	Ref      string //sequence 1 base(s)
+	Alt      string //sequence 2 base(s)
+}
+
+func usage() {
+	fmt.Print("locateCG chr*.fa.gz chromName \n" +
+		"\tRecords positional information of CpG site changes in sequence 1 relative\n" +
+		"\tto sequence 2 in alignment fasta. Only records changes resulting from substitutions, not indels.\n" +
+		"\tARGS:\n" +
+		"\tchr*fa.gz - input fasta file  with 2 aligned sequences to compare\n" +
+		"\tchromName - name of chromosome e.g. chr1\n")
+}
+
+func main() {
+	//expect 2 args: input fasta file and chromosome name
+	var expectedNumArgs int = 2
+
+	flag.Usage = usage
+	log.SetFlags(log.Ldate | log.Ltime | log.Lshortfile)
+	flag.Parse()
+
+	//if number of provided args is not 2
+	if len(flag.Args()) != expectedNumArgs {
+		flag.Usage()
+		log.Fatalf("Error: expecting %d arguments, but got %d\n", expectedNumArgs, len(flag.Args()))
+	}
+
+	//args
+	inFa := flag.Arg(0)
+	chrom := flag.Arg(1)
+
+	s := Settings{
+		InFa:      inFa,
+		ChromName: chrom,
+	}
+
+	//run locate() function on provided args; will output a slice of structs
+	locateCG(s)
+}
+
+func locateCG(s Settings) {
+	var err error
+	var output []CGsite
+	//set chromosome name from args
+	chrom := s.ChromName
+
+	//read in fasta file
+	f := fasta.Read(s.InFa)
+
+	//specify sequence 1 and 2
+	firstSeq := f[0].Seq
+	secondSeq := f[1].Seq
+
+	//if no sequences found
+	if len(firstSeq) == 0 || len(secondSeq) == 0 {
+		log.Fatalf("Missing or empty sequences for seq1 or seq2.")
+	}
+	//add check for equal lengths
+	if len(firstSeq) != len(secondSeq) {
+		log.Fatalf("Seq1 and seq2 not equal in length.")
+	}
+	//pre-allocate size of output
+	//output = make([]CGsite, 0, 10000)
+
+	//initialize pair of matched reference/alignment positions
+	refStart := 0
+	alnStart := 0
+
+	fmt.Printf("Beginning loop\n")
+	//for length of first sequence
+	for i := 0; i < len(firstSeq)-1; i++ {
+
+		//store dna.Bases
+		f1, f2 := firstSeq[i], firstSeq[i+1]
```

Comment:
you will get a minor memory efficiency gain if you initialize the f1 and f2 variable outside of the for loop similar to what you did in lines 67, 68. There are a few other examples of this below